### PR TITLE
CloudWatch: Fix passing of namespace for getting custom metrics #28075

### LINKED ISSUE #28075: Cloudwatch custom metric results in server error when query metricFindQuery
**What happened**:

Querying aws cloudwatch custom metrics results in a 500er server error when try to get the list of available metric names.

grafanabug

Server log 

```
/grafana.....de ecs/grafana/f05454fc888745d99c4f82f261266df8 t=2020-10-07T09:30:21+0000 lvl=eror msg="Request Completed" logger=context userId=2 orgId=1 uname=markuman method=POST path=/api/tsdb/query status=500 remote_addr=10.103.11.8 time_ms=27 size=34 referer="https://grafana......de/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22AWS%20CloudWatch%22,%7B%22namespace%22:%22fluentd_metrics%22,%22metricName%22:%22%22,%22expression%22:%22%22,%22dimensions%22:%7B%7D,%22region%22:%22default%22,%22id%22:%22%22,%22alias%22:%22%22,%22statistics%22:%5B%22Average%22%5D,%22period%22:%22%22,%22matchExact%22:true%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D"

```

Browser Payload

```json
{"from":"1602059279729","to":"1602062879731","queries":[{"refId":"metricFindQuery","intervalMs":1,"maxDataPoints":1,"datasourceId":2,"type":"metricFindQuery","subtype":"metrics","region":"eu-central-1","namespace":"fluentd_metrics"}]}
```

Browser response 

```json
{"message":"Metric request error"}
```

**What you expected to happen**:

Autocomplete existings metric names and its dimensions.  
When a dimension is set, the autocomplete works again for its value.

**How to reproduce it (as minimally and precisely as possible)**:

Try to access aws cloudwatch custom metric.

**Anything else we need to know?**:

It works in the past.  But I cannot say for sure in which version it failed.  
Out past used versions. I believe 7.0.x it works fine.  
```
7.0.0
7.0.3
7.0.6
7.1.3
```

**Environment**:
- Grafana version: 7.2.0
- Data source type & version: CloudWatch
- OS Grafana is installed on: Official Docker Container
- User OS & Browser: Firefox 81

### PR DESCRIPTION
**Pull request by mitchmckenzie**
**What this PR does / why we need it**:

This PR fixes a bug that omits the cloudwatch custom metrics namespace value from getting passed to the ListMetricsPages  api call.

**Which issue(s) this PR fixes**:

Fixes #28075

**Special notes for your reviewer**:

No modifications were made to the existing unit test as it already covers the lines that were modified in this PR.  The unit test didn't catch this problem since the aws cloudwatch api calls are mocked. I have functionally tested the changes locally and it does resolve the issue.

### CODE REVIEW

**Comment by aknuds1** [comment]
@papagian is already working on a fix for this issue.

**Review by papagian — COMMENTED** [review]
Thank you for contributing! Your changes LGTM; I've pushed some improvements so as the code is more readable:

- I've changed custom metrics and dimension cache from map[string]map[string]map[string]*customMetricsCache to map[string]*customMetricsCache
- I've tried to reduce as much as possible the number of times that we parse datasource JSON (by calling `getDSInfo`)

**Comment by papagian** [comment]
@mitchmckenzie Can you please sign the Contributor License Agreement? We would like to merge it as soon as possible!

@aknuds1 Would you like to review the modifications?

**Comment by aknuds1** [comment]
Thanks @papagian, I will have a look later when I have the time.

**Review by aknuds1 — COMMENTED** [review]
Generally LGTM, but I made some code improvements in a [branch](https://github.com/grafana/grafana/tree/bugfix/28075). If @papagian  and @wbrowne are OK with them, I can push them to this branch.

File: pkg/tsdb/cloudwatch/metric_find_query.go [hunk 1]
```diff
@@ -389,7 +400,9 @@ func (e *cloudWatchExecutor) handleGetMetrics(ctx context.Context, parameters *s
 	} else {
 		var err error
 		dsInfo := e.getDSInfo(region)
-		if namespaceMetrics, err = e.getMetricsForCustomMetrics(region, namespace, dsInfo.Profile); err != nil {
+
+		p := props{region: region, namespace: namespace, profile: dsInfo.Profile}
+		if namespaceMetrics, err = e.getMetricsForCustomMetrics(p); err != nil {
```
Comment (aknuds1) [inline pkg/tsdb/cloudwatch/metric_find_query.go:405]: This change doesn't improve anything, I think the changes in my branch should be applied, where `e.getDSInfo` is called by `e.getMetricsForCustomMetrics` itself. The reason is that with my changes, the burden on the caller is smaller and you get less complicated code.

File: pkg/tsdb/cloudwatch/metric_find_query.go [hunk 2]
```diff
@@ -775,12 +789,12 @@ func (e *cloudWatchExecutor) getMetricsForCustomMetrics(region, namespace, profi
...
 var dimensionsCacheLock sync.Mutex
 
-func (e *cloudWatchExecutor) getDimensionsForCustomMetrics(region, namespace, profile string) ([]string, error) {
+func (e *cloudWatchExecutor) getDimensionsForCustomMetrics(props props) ([]string, error) {
 	dimensionsCacheLock.Lock()
 	defer dimensionsCacheLock.Unlock()
 
 	bldr := strings.Builder{}
-	for i, s := range []string{profile, region, namespace} {
+	for i, s := range props.toArray() {
```
Comment (aknuds1) [inline pkg/tsdb/cloudwatch/metric_find_query.go:797]: To simplify here, the algorithm for computing the cache key should be extracted instead. But I would do that in another PR than this hotfix.

**Comment by papagian** [comment]
> Thank you for contributing! Your changes LGTM; I've pushed some improvements so as the code is more readable:
> 
> * I've changed custom metrics and dimension cache from map[string]map[string]map[string]*customMetricsCache to map[string]*customMetricsCache
> * I've tried to reduce as much as possible the number of times that we parse datasource JSON (by calling `getDSInfo`)

I have reverted my modifications. We want to fix this as soon as possible; we will address the additional issues in separate PR.

**Comment by aknuds1** [comment]
Closing this PR since the CLA isn't signed, and we're short on time to get the fix into v7.2.2. We made #28391 instead.

**Comment by mitchmckenzie** [comment]
@aknuds1 , I have signed the CLA 3 times. Not sure why it’s not getting linked to my account.  Strange. Thanks for getting the fix merged so quick.

**Comment by aknuds1** [comment]
> @aknuds1 , I have signed the CLA 3 times. Not sure why it’s not getting linked to my account. Strange. Thanks for getting the fix merged so quick.

Aha, well, that's weird :/
